### PR TITLE
[draft] adapter: distributed timestamp oracle backed by "postgres"

### PR DESCRIPTION
**Pull request by aljoscha**
### Motivation

A part of #22029 , but so far a draft for shopping the thing around.

The commits tell the story of the evolution, with intermediary steps, like the consensus-backed timestamp oracle that can be removed again. Also, there's more steps we could do like now share the final version of the postgres oracle using an `Arc<dyn TimestampOracle>`, meaning we wouldn't have to do the shallow clones anymore.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/coord/timestamp_oracle.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+pub mod postgres;
+
+/// A type that provides write and read timestamps, reads observe exactly their
+/// preceding writes.
+///
+/// Specifically, all read timestamps will be greater or equal to all previously
+/// reported completed write timestamps, and strictly less than all subsequently
+/// emitted write timestamps.
+#[async_trait]
+pub trait TimestampOracle<T> {
```
Comment (danhhz) [inline src/adapter/src/coord/timestamp_oracle.rs:31]: can this whole module be its own crate? if so, i'd lean toward do it

File: src/adapter/tests/timestamp_selection.rs [hunk 2]
```diff
@@ -261,6 +264,8 @@ fn parse_query_when(s: &str) -> QueryWhen {
 /// Transaction isolation can also be set. The `determine` directive runs determine_timestamp and
 /// returns the chosen timestamp. Append `full` as an argument to it to see the entire
 /// TimestampDetermination.
+// allow `futures::block_on` for testing.
```
Comment (danhhz) [inline src/adapter/tests/timestamp_selection.rs:267]: this turns out to be a pretty big footgun in practice, even for tests. highly recommend figuring out a way to use tokio's block_on

File: src/adapter/src/coord/timestamp_oracle.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+//! exactly their preceding writes.
+//!
+//! Specifically, all read timestamps will be greater or equal to all previously reported completed
+//! write timestamps, and strictly less than all subsequently emitted write timestamps.
+
+use async_trait::async_trait;
+
+use crate::coord::timeline::WriteTimestamp;
+
+pub mod consensus;
```
Comment (danhhz) [inline src/adapter/src/coord/timestamp_oracle.rs:20]: are we planning on merging all three impls or just the final one?

File: src/adapter/src/coord/timestamp_oracle.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+pub mod durable;
```
Comment (danhhz) [inline src/adapter/src/coord/timestamp_oracle.rs:21]: maybe a better name for this than "durable" is "catalog"? they're all durable, no?

File: src/adapter/src/coord/message_handler.rs [hunk 5]
```diff
@@ -179,10 +183,10 @@ impl Coordinator {
         if let Err(err) = self.catalog_transact(None, ops).await {
             tracing::warn!("Failed to update storage metrics: {:?}", err);
         }
-        self.schedule_storage_usage_collection();
+        self.schedule_storage_usage_collection().await;
     }
 
-    pub fn schedule_storage_usage_collection(&self) {
+    pub async fn schedule_storage_usage_collection(&self) {
```
Comment (danhhz) [inline src/adapter/src/coord/message_handler.rs:189]: might be nice to keep this non-async so it's obvious that it doesn't do anything expensive. which I think could be as easy as passing in the ts as an arg?

**Review by danhhz — COMMENTED** [review]
left a few superficial comments from a very high-level scan, don't bother doing anything in response to them, i'm mostly flushing them out so github doesn't eat them. i'll do a deeper pass later this week

one high-level thing i noticed is that this makes quite a few non-async fns into async fns. from what I understand, it's possible that those were intentionally non-async so it's easier to reason about what's happening in the coord loop. we should check with folks on that

File: src/adapter/src/coord/timestamp_oracle/postgres.rs [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+        read_ts
+    }
+
+    #[tracing::instrument(name = "oracle::apply_write", level = "debug", skip_all)]
+    async fn apply_write(&mut self, write_ts: Timestamp) {
+        let q = r#"
+            UPDATE timestamp_oracle SET write_ts = GREATEST(write_ts, $2), read_ts = GREATEST(read_ts, $2)
+                WHERE timeline = $1;
+        "#;
+        let client = self.get_connection().await.expect("todo");
```
Comment (philip-stoev) [inline src/adapter/src/coord/timestamp_oracle/postgres.rs:327]: This is being seen in Sentry - https://materializeinc.sentry.io/issues/4478589600/?project=6780145&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=7d&stream_index=10

**Comment by philip-stoev** [comment]
@aljoscha this looks like something that could use some more chaos testing and such. Please ping the QA team and/or add me as a reviewer once the time arrives . Thank you!

**Review by jkosh44 — COMMENTED** [review]
This looks good to me, I didn't have many useful comments.

**Comment by philip-stoev** [comment]
@aljoscha please re-request a review from the QA team once the merge conflicts are resolved. I would like to run another Nightly at the very least.

**Comment by aljoscha** [comment]
#22262 has been merged

**Pull request closed without merging**